### PR TITLE
phrases in conversations

### PR DESCRIPTION
**Pull request by UnorderedSigh**
**Feature:** This PR implements the feature request detailed and discussed in issue #7322

Also, it may be a viable alternative to the feature requested in #7847

## Feature Details
Allows phrases in conversations:

```bash
mission "My Mission"
  ...
  on offer
    conversation
       `"Before, you could only have static text," said Captain <last>`
       `    "Now you can include ${phrases} inside your conversations!" ${Dr. First Last} exclaims.`

phrase "Dr. First Last"
    word
        Dr.
        Ms.
        Mr.
    word
        "<first> <last>"
```

The phrase is expanded when the mission is instantiated. Also, the phrases can contain substitutions, like `<first> <last>`. See the "Testing Done" for an example.

There is a new `Phrase::ExpandPhrases()` which should make it easy to put phrases anywhere.

## UI Screenshots

Generated by the test case in the Testing Done section. This is a composite of three screenshots, from the three different parts of the conversation.

phrases-in-conversations

## Usage Examples
Randomly-generated missions can have conversations instead of just dialogs.

## Testing Done

1. Put this in data/[phrasetest.txt](https://github.com/endless-sky/endless-sky/files/10470686/phrasetest.txt)
2. Depart
3. Land
4. Visit spaceport

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

File: source/Conversation.cpp [hunk 1]
```diff
@@ -186,20 +188,26 @@ void Conversation::Load(const DataNode &node, const string &missionName)
...
 		}
 		// Check for common errors such as indenting a goto incorrectly:
-		else if(child.Size() > 1)
+		else if(child.Size() > 2 || (child.Size() == 2 && child.Token(0) != "phrase"))
 			child.PrintTrace("Error: Conversation text should be a single token:");
 		else
 		{
-			// This is just an ordinary text node.
+			// This is an ordinary text node (isPhrase=0) or a phrase (isPhrase=1)
+			int isPhrase = child.Size() == 2;
```
Comment (Hurleveur) [inline source/Conversation.cpp:196]:
```suggestion
			bool isPhrase = child.Size() == 2;
```

File: source/Conversation.cpp [hunk 2]
```diff
@@ -361,7 +369,17 @@ Conversation Conversation::Instantiate(map<string, string> &subs, int jumps, int
...
 	{
 		for(Element &element : node.elements)
+		{
+			if(element.isPhrase)
+			{
+				const Phrase * phrase = GameData::Phrases().Find(element.text);
+				if(phrase)
+					element.text = phrase->Get();
+				element.text += "\n";
+				element.isPhrase = false;
```
Comment (Hurleveur) [inline source/Conversation.cpp:379]: why?

**Comment by UnorderedSigh** [comment]
Pinging @Zitchas and @Hurleveur whose issues are listed in the PR description.

**Review by Hurleveur — COMMENTED** [review]
not sure if this is the best way to do this; containing the name of the phrase in text may be easier but I dont think it makes sense, it should probably be a new member of the elements

a simple Phrase *phrase = nullptr would do?

File: source/Conversation.cpp (hunk 2 above)
Comment (UnorderedSigh) [reply source/Conversation.cpp:379]: Why what?

**Comment by UnorderedSigh** [comment]
> not sure if this is the best way to do this; containing the name of the phrase in text may be easier but I dont think it makes sense, it should probably be a new member of the elements

Doing this now. I think it'll simplify the code changes.

File: source/Conversation.cpp (hunk 2 above)
Comment (Hurleveur) [reply source/Conversation.cpp:379]: if we keep this syntax we should rename isPhrase to holdsPhrase or smth like that; its still a phrase that's contained and idk why we would mark it as not a phrase all of a sudden

**Comment by RisingLeaf** [comment]
Hmmm, I like the idea, but the implementation should be different imo. What I would like to see is to use a phrase similar to a substitution in a conversation text.
Something like
`guy talking: "<phrase: generic guy talking phrase>"`

this would make this more flexible and I dont think it is too hard to do as text is already checked for substitutions you should be able to just add a check there.

**Comment by UnorderedSigh** [comment]
> this would make this more flexible and I dont think it is too hard to do as text is already checked for substitutions you should be able to just add a check there.

It would need to be recursive so that a `<first> <last>` inside a phrase is expanded.

**Comment by Hurleveur** [comment]
If we do want to do smth like that we may as well allow putting phrases into substitutions themselves?

**Comment by RisingLeaf** [comment]
> > this would make this more flexible and I dont think it is too hard to do as text is already checked for substitutions you should be able to just add a check there.
> 
> It would need to be recursive so that a `<first> <last>` inside a phrase is expanded.

not sure if I understand what you are saying.

**Comment by UnorderedSigh** [comment]
There are two problems with "guy says: <phrase: blah>"

1. The `<whatever>` is handled inside Format::Replace, which receives a `map<string,string>` for the replacements. 
2. If phrase "blah" includes `<first> <last>` or other subs, they won't be expanded.

A simpler method may be:

```
    `A guy says: ${blah}`
```

Which would be handled like this:

1. If there is a `$` then expand phrases in those `$`
2. Expand the `<subs>`

That way you can have `<subs>` in phrases and text, in an unambiguous manner, with few code changes.

**Comment by RisingLeaf** [comment]
> There are two problems with "guy says: <phrase: blah>"
> 
>     1. The `<whatever>` is handled inside Format::Replace, which receives a `map<string,string>` for the replacements.
> 
>     2. If phrase "blah" includes `<first> <last>` or other subs, they won't be expanded.
> 
> 
> A simpler method may be:
> 
> ```
>     `A guy says: ${blah}`
> ```
> 
> Which would be handled like this:
> 
>     1. If there is a `$` then expand phrases in those `$`
> 
>     2. Expand the `<subs>`
> 
> 
> That way you can have `<subs>` in phrases and text, in an unambiguous manner, with few code changes.

This looks good, if you can do that, that would be nice.

**Comment by UnorderedSigh** [comment]
> This looks good, if you can do that, that would be nice.

I can. The changes will be slightly longer in line count since I'll split up Phrase::Choice::Choice, but the result will be less complex.

**Comment by UnorderedSigh** [comment]
The deed is done. You can now put `${phrases}` inside mission text, and I added GameData::ExpandPhrases to expand them.

**Review by warp-core — COMMENTED** [review]
Any particular reason you've chosen to put the new method in GameData?
I think it should either go in a new class (probably in the text folder) or in Phrase itself, or in Format.
I note that it appears to reproduce some logic from the constructor of Phrase::Choice, so perhaps it should go in Phrase and the repeated code could be moved to a namespace method?

**Comment by UnorderedSigh** [comment]
> Any particular reason you've chosen to put the new method in GameData? I think it should either go in a new class (probably in the text folder) or in Phrase itself, or in Format. I note that it appears to reproduce some logic from the constructor of Phrase::Choice, so perhaps it should go in Phrase and the repeated code could be moved to a namespace method?

Ah, now you feel the confusion I felt. These are the options I considered and why I did or did not use them:

1. Conversation - putting it here would be neat and clean, but not make it generally available.
2. Somewhere in text/ - this code looks like it is intended to be entirely independent of the rest of Endless Sky, especially the GameData.
3. Phrase static - The Phrase class doesn't know where the unique list of game data phrases reside
4. Phrase member - Of course not. You don't have a Phrase to start with.
5. GameData - This is where the phrases reside, so there's no ambiguity.
6. New file - Seems over-complicated to add a new file just for one function.

Given that you're specifically using the GameData::Phrase()s to replace the strings, GameData seems the best place to me.

**Comment by warp-core** [comment]
> Phrase static - The Phrase class doesn't know where the unique list of game data phrases reside

I'm not quite sure what you mean by this.
`Phrase.cpp` includes `GameData.h` and already has code that retrieves a Phrase pointer ffrom GameData::Phrases().

https://github.com/endless-sky/endless-sky/blob/2e53db35cd765d31030eb79ce53b750011a9fd47/source/Phrase.cpp#L115-L159

**Comment by UnorderedSigh** [comment]
> reproduce some logic from the constructor of Phrase::Choice, so perhaps it should go in Phrase and the repeated code could be moved to a namespace method

I tried this, but the differences are substantial enough that the resulting function would be a mess. Notice that one of them is assembling a phrase in member variables while the other is appending GameData::Phrase()s to string. The code is far less similar than it appears.

**Comment by UnorderedSigh** [comment]
>  Phrase.cpp includes GameData.h and already has code that retrieves a Phrase pointer ffrom GameData::Phrases().

Sounds good. I'll move it to a static Phrase::ExpandPhrases()

**Comment by UnorderedSigh** [comment]
Done! We now have Phrase::ExpandPhrases().

**Review by Hurleveur — COMMENTED** [review]
This seems very similar to Format::Replace. How about making two Replace functions, one with the list of substitutions, the other one without
Both would call an internal function that would take the separating strings ("<" and ">" or "${" and "}") and a bool to know if it should do phrase/substitution replacement?

edit: if thats too much of a mess I think this could be fine as is, but doing it this way would make formatting more powerful?

**Comment by UnorderedSigh** [comment]
> edit: if thats too much of a mess I think this could be fine as is, but doing it this way would make formatting more powerful?

It will make the PR much longer, and the text/ directory would no longer be separate from the rest of EndlessSky. That would seem to break the design.

However, if I get more requests for it, I can implement that. I'd rather do it in a later PR though.

**Comment by warp-core** [comment]
> > edit: if thats too much of a mess I think this could be fine as is, but doing it this way would make formatting more powerful?
> 
> It will make the PR much longer, and the text/ directory would no longer be separate from the rest of EndlessSky. That would seem to break the design.
> 
> However, if I get more requests for it, I can implement that. I'd rather do it in a later PR though.

While my original suggestion did include putting this code in Format or some other `text/` class, your comments and looking at the code have convinced me that that isn't actually a desirable change.

While this is somewhat similar to what Format::Replace() does, the differences mean unifying the two would almost certainly result in an increase in complexity and reduce maintainability, not improve it.

**Comment by UnorderedSigh** [comment]
I plan on allowing conditions in phrases. That, I think, is the point at which we should consider the complexity of unifying the subs, phrase, and condition expansion into a single function or class. For now, in my opinion, it is okay to separate them.

**Review by Hurleveur — APPROVED** [review]
pretty simple code wise and I agree with how it was done in terms of doing a new function for it

speaking as a bit of a CC this is a really big improvement

**Review by RisingLeaf — APPROVED** [review]
nice and clean

File: source/Phrase.h [hunk 3]
```diff
@@ -30,6 +30,10 @@ class DataNode;
 // Class representing a set of rules for generating text strings from words.
 class Phrase {
 public:
+	// Replace all occurrences ${phrase name} with the expanded phrase from GameData::Phrases()
+	static std::string ExpandPhrases(const std::string &source);
+
+
```
Comment (warp-core) [inline source/Phrase.h:36]:
```suggestion

public:
```

**Review by warp-core — APPROVED** [review]

**Pull request merged**